### ISSUE TITLE
[Question] Is it possible to hide secrets from trace viewer?

### ISSUE THREAD

**Issue by ValorHeart** [microsoft/playwright#19992]
While logging into my application, the input field contains a password which is a secret that I prefer to hide, is it possible to hide it in the trace viewer?
<img width="475" alt="image" src="https://user-images.githubusercontent.com/61405630/211558455-b26d201d-80cf-43ea-b533-90ed4b8e83f0.png">

**Comment by MindaugasMateika** [comment]
This is not the first issue regarding this issue and I would very much need such feature myself. An option for fill and type to mask the inputted value from reports and traces would be a life saver. Something like:

```
await someLocator.fill('secret text', { maskInReport: true })
```

**Comment by aslushnikov** [comment]
> While logging into my application, the input field contains a password which is a secret that I prefer to hide, is it possible to hide it in the trace viewer?

@ValorHeart there's no way to do so. Even if we mask if from the step itself, we cannot guarantee that
- there's no DOM element that holds the sensitive value (e.g. form input value)
- there's no HTTP request that carried the sensitive value over
- there's no console message that dumped the value

and so on. Traces also store login cookies that can be used to login to the service.

All in all, filtering traces for sensitive information is an uphill battle. I'd recommend against sharing sensitive traces with untrusted parties.

**Comment by MindaugasMateika** [comment]
The mentioned cases are out of scope for Playwright but what is within the scope is to not leak sensitive data in .fill or .type logs. It's a bit dissapointis that these issues are simply closed (this is not the first) because it would certainly get upvotes and is needed by community.

**Comment by chuanqisun** [comment]
Agree with @MindaugasMateika

Just an idea: grep all environment variable (by their values) and remove them from the output folder. It isn't too hard to implement and will make the trace sharing more secure for everyone. Please re-consider this feature request.

**Comment by carlnoval** [comment]
Posted about something similar in discussion: https://github.com/microsoft/playwright/discussions/20902 before finding this thread. Though implementation would be an uphill battle, this feature would greatly help with security especially when sharing reports with steak holders.

Please re-consider this feature request.

**Comment by jpinheiro27** [comment]
This feature should be a must have in terms of security, please could you re-consider it?

**Comment by p01** [comment]
> All in all, filtering traces for sensitive information is an uphill battle. I'd recommend against sharing sensitive traces with untrusted parties.

We don't even have to go as far as sharing traces featuring passwords for them to be problematic.

Traces featuring credentials will trigger automated compliance checks that can invalidate the builds and trigger follow up security audits.

Your points about the credentials being potentially replicated throughout the DOM, or that Github actions filter them out are valid, but it's not a reason enough to dismiss the issue. 

Playwright should have a boolean config to scrub logs and traces before they are saved on disk, or emitted to std.
It could even be a little bit smarter and collect the credentials in the expected places, then do another pass to scrub them from any other place to avoid the case you mentioned about credentials being sprinkled in the DOM.

**Comment by OlllllllO** [comment]
CircleCI also obscures passwords when they print out logs so it is possible. Sure, there are always methods to get around it but exposing a password should not be the default.

ps - what does the `P3-collecting-feedback` label mean?

**Comment by ValorHeart** [comment]
@OlllllllO It means that it is priority 3, and they are waiting to see if there are upvotes (thumbs up on the original message) in order consider developing solution

**Comment by lux-capacitor** [comment]
@ValorHeart  This is a complete security flaw - What's the purpose of even using an Environment Variable if you print all sensitive keys into the literal test steps for anyone to see in the application? Considering Checkly was hacked before and had to issue public statements on the last breach, I'm impressed to see a continued nonchalant attitude towards such a heinous security issue.

**Comment by jwohlfahrt** [comment]
I could also really use this feature, as I'm having to authenticate my app under test via Okta, and we can't afford to expose credentials in our html reports, which we make available across our qa organization.

**Comment by Jerome1337** [comment]
We need this feature to. Every logging credentials are displayed into the execution steps. Should be great to be able to deactivate manually traces from output.

**Comment by Wassim24** [comment]
It would be a great-added value if we had this ability to hide sensitive information.

**Comment by OS-stanislavivanov** [comment]
To "Mask sensitive data in html report" it's important feature when:
- you're working with multiple teams that may/should not know about the others team secrets
- you're having a common pipeline to trigger the tests for different teams when playwright project configurations are following the same guidelines.
- you're having a common standard for all teams that are creating tests using Playwright working as isolated units in the same repo.
I believe there are way more people that are gonna be thankful for this feature then in this thread 😄 
So please take this in consideration.

**Comment by myieye** [comment]
FWIW, it's not very hard to mask traces yourself. E.g. here's a powershell script that masks all the environment variables that start with `PLAYWRIGHT_SECRET_`: (e.g. `./mask-sript.ps --traceDir ./path-to-your/traces`)
```ps
param(
    [Parameter(Mandatory)][string] $traceDir
)
Add-Type -Assembly  System.IO.Compression.FileSystem

# trace = Core trace info
# network = Network log
# json = Network log payloads
$filesToMask = "\.(json|trace|network)$"
$files = Get-ChildItem -path $traceDir -Filter *.zip
$secrets = Get-ChildItem env:PLAYWRIGHT_SECRET_*

Write-Output "Masking $($secrets.Count) secrets in $($files.Count) traces.";

foreach ($file in $files) {
    try {
        Write-Output "Masking $file"
        $zip = [System.IO.Compression.ZipFile]::Open($file.FullName, "Update")
        $entries = $zip.Entries.Where({ $_.Name -match $filesToMask })
        $traceMaskCount = 0;
        foreach ($entry in $entries) {
            $reader = [System.IO.StreamReader]::new($entry.Open())
            $content = $reader.ReadToEnd()
            $entryMaskCount = 0;
            foreach ($secret in $secrets) {
                $pieces = $content -split "\b$($secret.Value)\b"
                $entryMaskCount += $pieces.Count - 1;
                $content = $pieces -join "*******"
            }
            $reader.Dispose()
            $writer = [System.IO.StreamWriter]::new($entry.Open())
            $writer.BaseStream.SetLength(0)
            $writer.Write($content)
            $writer.Dispose()
            Write-Output "- $entry ($entryMaskCount)"
        }
        Write-Output "Finished masking $file ($traceMaskCount)"
    }
    catch {
        Write-Warning $_.Exception.Message
        continue
    }
    finally {
        if ($zip) {
            $zip.Dispose()
        }
    }
}
```

But....for us, the goal was to not leak sensitive data when uploading the traces as GitHub workflow artifacts. So, masking was not the right solution at all due to all the reasons that @aslushnikov [mentioned](https://github.com/microsoft/playwright/issues/19992#issuecomment-1378027355). Instead we just started zipping them up with a password:
```
- name: Password protect Playwright traces
  if: always()
  shell: bash
  env:
    ZIP_PASSWORD: ${{ secrets.TEST_USER_PASSWORD }}
  run: 7z a ./playwright-traces.7z ./bin/playwright-traces/*.* -p"$ZIP_PASSWORD"
```

**Comment by nitzanshuker** [comment]
Looking for something like this for a long time

**Comment by sebast-c** [comment]
This is an important feature that we also need to be able to exploit trace mode

**Comment by mikicymulate** [comment]
@yury-s 
Is there a chance this will happen in the near future?

**Comment by alexkuc** [comment]
While not the most elegant solution, I ended up doing the following:

- GitHub Action output is masked using [`.setSecret`](https://github.com/actions/toolkit/tree/main/packages/core#setting-a-secret)
- Actual artifact (trace & report) are tarballed and encrypted using GPG; if you set up the GPG password as organization secret, you can re-use it across many repositories

**Comment by yury-s** [comment]
We've evaluated various approaches to filtering secrets of the traces but all of them are heuristics/best effort approaches which we won't be able to recommend to our clients as a generic solution. Basically, the application code for which we record the trace can be arbitrarily complex and log the original password in altered forms (e.g. reversed, or transformed via "cipher disk" etc.) and we won't detect that. If this requirement can be relaxed, one can postprocess the trace via custom script or follow approach from [the comment above](https://github.com/microsoft/playwright/issues/19992#issuecomment-1869449848).

Given the range of potential pitfalls, our advice is to safeguard trace access with the same rigor applied to password protection and not share it publicly.

**Comment by ZooDoo4U** [comment]
Might suggest the suggestion i put forward solves most if not all the issue, never exposes the sensitive data, even better throws all the responsibility on the PW user where it belongs and gets the PW out of the crosshairs...  As the one writing the tests i should know which fields are sensitive and would never assume PW would ever have a clue...  Might be setting a users login, and from a security perspective would prefer both user/login name protected as well as the password..   Getting the user login, just took a large number of invalid logins away...  In general i would like the ability to mask out any value(s) to be protected, not just password fields,  could be entering HIPAA or any other sensitive values.   And the problem with the post processing, now need to worry about data in transition and a secondary script/program to wipe the values, and other processing issue, i.e. file might become read only or the program running gets downgraded permissions.  Just saying throwing the burden of a separate program isn't secure at all.  

Not sure of the resistance of having either a fillSecure or fill( {options}) to prevent the string from being printed...

**Comment by RachBreeze** [comment]
I am currently comparing Playwright and Cypress. Playwright, due to it being backed by Microsoft, its logs and screenshots were winning, but to find out that it doesn't mask inputs or other GDPR issues saddens me when we're looking to create GDPR-compliant tests. To add, even if we use test user details and accept that our work IP is exposed, in some scenarios, we will be entering live credit card data, which, if breached, is a huge security risk. Other tools developed by Microsoft such as Microsoft Clarity do obsfucate this data I don't understand the resistance, especially if other testing tools (such as Cypress and BrowserStack do this).

**Comment by RachBreeze** [comment]
Hello

I have been evaluating Playwright and Cypress. I really like that Playwright is backed by Microsoft and up until I found out that it wasn't possible to hide secrets from trace viewer on test failure, Playwright was scoring more in the comparisons.

I am a bit surprised at the apparent unwillingness to resolve the issue. There have been some great suggestions on this ticket as to resolve it. It does leave a bit of a GDPR and security hole for it. For example there are some situations where test data won't cut it as a testing tool, for example in the case of credit card details where we need to test the integration on a live environment.

We can work hard to ensure that credit card details and other sensitive data are hidden by putting them in environment variables, but the minute they are in a failed test, they can be captured; makeing the hard work of obscuring the sensitive data pointless. It means for example if credit card details are captured someone can go on a spending spree. 

I have Microsoft Clarity installed on my own website and can see this redacts sensitive information by default in recordings, so I know that Microsoft can redact information without any guidance from a tester/ QA. It would make sense to add this 

Whilst @aslushnikov argues for "only share the traces with people you trust" I would prefer to argue for security and GDPR compliance by obscurity.  We know that Cypress, Browser Stack and Microsoft Clarity can do this, so I struggle to see the issue, with adding this to Playwright .

**Comment by modern-sapien** [comment]
I wonder how difficult it would be to create a 3rd party package that handles this.

**Comment by fungairino** [comment]
This is a really big miss on Playwright's part. At the very least some big red flag around this issue should be shown on the environment variables section as well as the authentication sections of the documentation.
https://playwright.dev/docs/test-parameterize#passing-environment-variables
https://playwright.dev/docs/auth

We recently accidentally leaked our test user account's google password due to this exact issue following Playwright's recommendations and best practices.

**Comment by garretlowe** [comment]
A potential workaround, for those interested, would be to set up your authentication inside of a [Fixture](https://playwright.dev/docs/test-fixtures).

This will cause your trace to no longer log any of the steps up to and including authentication. At least not in enough detail to show unencrypted credentials.

**Comment by samixchoumi** [comment]
Hello, any update on this subject ? Because regarding ISO-27001, it should be a must have !

**Comment by dparadisZentelia** [comment]
The way I found, is to use another project for the login and set the trace off it.
The project setup (where the login was made) have the trace set to off and the other project chrome-en have a dependencie on it.
I know, this is not perfect, because the token is visible in the other trace but that prevent to have a clear password in the trace.
Hope that will help you!

In your playwright.config.ts :
```
projects: [
    // Setup project use to login on SSO
    { name: 'setup', 
      testMatch: /.*\.setup\.ts/,
      //to be sure no password will be keep on trace
      use : {trace : 'off'}
    },

    {
      name: 'chrome-en',
      use: { ...devices['Desktop Chrome'],
        locale: 'en',
        storageState: 'playwright/.auth/currentUser.json',
        trace: 'on',
        testIdAttribute: 'testid', 
        viewport: { width: 1600, height: 900 },
        actionTimeout: 10000,
       },
      //force SSO login
      dependencies: ['setup'],
    }]
```
![no_trace](https://github.com/user-attachments/assets/204135e3-6bd9-4a85-b987-620bfd2250a0)

**Comment by Guzikowski** [comment]
We are looking for a solution to not expose our sensitive information in the trace logs as well. In most cases our test data is sanitized to the point it is no longer considered personal identification information. Where things are at risk is when we call our authentication service or our token provider. We understand that this is difficult, but could we omit fields or omit traffic recorded to certain URLs. Ideally would like to have a simple masking of identified fields in the network traffic. 

Currently, we would need to turn off all recording of artifacts to test in our more sensitive environments like production and UAT environments, as well as keep these tests in a separate repository to not accidentally expose them to recording.

**Comment by pubbalk** [comment]
Was able to hide password in trace, by Intercepting the request data and replacing filled value in the textbox with actual password value. Just sharing, hope it will useful for someone.

```
await this.passwordTextbox.fill('SomePasswordToInterceptToHideInTrace'); // This value will display on the trace, so put any value, not the real secret/password.

const password = 'RealPassword';

// Start Intercepting the requests, find and replace the fake password with real password
await this.page.route('**/*', (route, request) => {
    if (request.method() === 'POST') {
        let modifiedPostData = request.postData();
        if (modifiedPostData) {
            modifiedPostData = modifiedPostData.replaceAll('SomeEmailPasswordToInterceptToHideInTrace', password);
        }
        route.continue({ postData: modifiedPostData });
    } else {
        route.continue();
    }
});

await this.signInButton.click();
.......
.......
.......
await this.page.unroute('**/*'); //Close the intercepting

```

**Comment by joe-op** [comment]
This seems in line with masking a password in an input box.  It prevents a common issue such as someone looking over your shoulder and seeing your password while you enter it.  It doesn't prevent someone from looking over your shoulder while you open the network tools and inspect the request body.

I'd like to demo Playwright to some non-technical people, but having passwords displayed prominently in the trace, even test passwords for throwaway test accounts, looks pretty bad.

**Comment by ruubytes** [comment]
> This seems in line with masking a password in an input box. It prevents a common issue such as someone looking over your shoulder and seeing your password while you enter it. It doesn't prevent someone from looking over your shoulder while you open the network tools and inspect the request body.
> 
> I'd like to demo Playwright to some non-technical people, but having passwords displayed prominently in the trace, even test passwords for throwaway test accounts, looks pretty bad.

The go-to method right now is to write an obfuscation function and run it in an onExit() reporter hook, where you unzip the trace archives and replace all the secrets via this obfuscation function, then zip the traces back up, also replace any mention of the passwords in the html report.

**Comment by joe-op** [comment]
> The go-to method right now is to write an obfuscation function and run it in an onExit() reporter hook, where you unzip the trace archives and replace all the secrets via this obfuscation function, then zip the traces back up, also replace any mention of the passwords in the html report.

Thanks.  That's useful, but I should have clarified that I'm particularly thinking of UI mode.  It's a nice way to demo Playwright and give an idea of the kinds of things that are possible with it, but I don't feel super comfortable about the fact that secrets are visible in the UI mode "Actions" list.  I'm not considering a scenario where traces are being shared with untrusted users.  It's more that displaying secrets in the UI just makes Playwright look bad.  People will not be comfortable seeing that, even if they're trusted users who have access to the secrets in question.

**Comment by dacloutier-logmein** [comment]
Our security folks here have identified this as a threat vector.

**Comment by vinzee** [comment]
Just wanted to point out that we were able to address this by using CDP (Chrome DevTools Protocol) as follows:

```js
export async function fillSensitive(page: Page, element: ElementHandle, text: string): Promise<void> {
  await element.focus();
  const cdpSession = await page.context().newCDPSession(page);
  await cdpSession.send('Input.dispatchKeyEvent', { type: 'keyDown', key: 'a', commands: ['selectAll'] });
  await cdpSession.send('Input.dispatchKeyEvent', { type: 'keyUp', key: 'a' });
  await cdpSession.send('Input.insertText', { text });
  await cdpSession.detach();
}
``` 

This skips the password fill from being tracked in the Playwright Trace Viewer (Actions Tab).

**Comment by aragossa** [comment]
the "unzip, replace, zip back" approach comes up a few times in this thread, so i measured what it actually does.

built a throwaway login test, seeded 9 kinds of secrets, recorded a trace: 35 occurrences across 6 files inside the zip, plus the session cookie in storageState. trace viewer renders `Fill "Hunter2-Staging-Pass"` in plain text.

piping every text file through a redactor gets it to 9 occurrences and breaks the trace. `"wallTime":1789811509873` becomes `"wallTime":[HIDDEN:6893df]`, a bare token where a number was, so the line stops being valid json. pageId/callId/frameId/type get redacted too, they look like high entropy strings. viewer just says "could not load trace".

what worked: parse the jsonl, redact string values only, keep an allowlist for the structural fields. 35 occurrences down to 3, 111/111 lines still valid json, viewer opens and shows `Fill "[HIDDEN:b2a02c]"`. the 3 left were a uk phone number and an nhs number, my detector has no uk rules.

(edited: dropped a link to my own page, this thread wasn't the place for it.)

**Comment by d-koppenhagen** [comment]
Nice try @aragossa . I reported this.

**Comment by aragossa** [comment]
fair. removed the link, the measurements are the part that belongs here.

**Comment by PsiKai** [comment]
> [Adi Gal (@ValorHeart)](https://github.com/ValorHeart) there's no way to do so. Even if we mask if from the step itself, we cannot guarantee that
> 
> * there's no DOM element that holds the sensitive value (e.g. form input value)
> * there's no HTTP request that carried the sensitive value over
> * there's no console message that dumped the value
> 
> and so on. Traces also store login cookies that can be used to login to the service.
> 
> All in all, filtering traces for sensitive information is an uphill battle. I'd recommend against sharing sensitive traces with untrusted parties.

All we are asking for is a best effort approach on the `fill` method.  You don't have to cover every case.  The other cases mentioned are in the control of the test writers, the playwright trace logs are not...
